TA-puppet-alert-actions/bin/ta_puppet_alert_actions/puppet_plan_action.py

```diff
@@ -0,0 +1,130 @@
+# alert['global']['bolt_user_pass'] = puppet_bolt_user_pass
+# alert['global']['puppet_bolt_server'] = notnone(puppet_enterprise_console, puppet_bolt_server, helper)
+# alert['global']['puppet_action_hec_token'] = notnone(splunk_hec_token, puppet_action_hec_token, helper)
+# alert['global']['timeout'] = timeout
+
+# # Load the alert specific settings that are really the plan we're running
+# alert['param']['bolt_target'] = bolt_target
+# alert['param']['plan_name'] = plan_name
+# alert['param']['plan_parameters'] = plan_parameters
+# alert['param']['puppet_environment'] = puppet_environment
+
+
+def run_puppet_plan(alert, helper):
+  # load our URLs, we generate possible ones assuming the console hostname is valid
+  # however if a user provides their own pdb or bolt url it goes here
+  # this also allows for us to add an int_proxy feature in the future
+  endpoints = pie.util.getendpoints(alert['global']['puppet_enterprise_console'])
+  rbac_url = endpoints['rbac']
+
+  # since we're swapping out this data upstream anyway
+  bolt_endpoint = pie.util.getendpoints(alert['global']['puppet_bolt_server'])
+  bolt_url = bolt_endpoint['bolt']
+
+  puppet_environment = alert['param']['puppet_environment']
+  splunk_hec_url = alert['global']['splunk_hec_url']
+  puppet_action_hec_token = alert['global']['puppet_action_hec_token']
+  plan_name = alert['param']['plan_name']
+  plan_parameters = alert['param']['plan_parameters']
+
+  # parse if we were given a hostname already
+  if alert['global']['pe_console'] is not None and alert['global']['pe_console'] is not '':
+    pe_console = alert['global']['pe_console']
+  else:
+    pe_console = endpoints['console_hostname']
+
+  message = {
+    'message': 'Running plan {} on {} '.format(plan_name,pe_console),
+    'action_type': 'plan',
+    'action_name': plan_name,
+    'action_parameters': plan_parameters,
+    'action_state': 'starting',
+    'pe_console': pe_console,
+  }
+
+  helper.log_debug(message)
+  pie.hec.post_action(message, pe_console, splunk_hec_url, puppet_action_hec_token)
+
+  bolt_user = alert['global']['bolt_user']
+  bolt_user_pass = alert['global']['bolt_user_pass']
+
+  if alert['global']['timeout'] is not None and alert['global']['timeout'] is not '':
+    plan_timeout = alert['global']['timeout']
+  else:
+    plan_timeout = 600
+
+  try:
+    token_lifetime = int(plan_timeout) * 2
+  except Exception as e:
+    helper.log_error("Timeout must be an integer, '{}' was provided instead".format(plan_timeout))
+
+ # Check if the user is configured with an RBAC token or Password:
+  if alert['global']['pe_token']:
+    auth_token = bolt_user_pass
+  else:
+    auth_token = pie.rbac.genauthtoken(bolt_user,bolt_user_pass,'TA-puppet-alert-actions',rbac_url, timeout=token_lifetime)
+
+  # note: parameters is expected as a text string, not json, so in sample alert json must be represented as:
+  # "plan_parameters": "{ \"target\": \"puppet.angrydome.org\", \"service\": \"puppet\"}"
```

Comment:
what's an angrydome?

Reply:
Thats probably an old testing hostname. I went ahead and updated this with a generic example hostname.